src/client/datascience/notebook/statusBarProvider.ts

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+'use strict';
+
+import { injectable } from 'inversify';
+import {
+    CancellationToken,
+    NotebookCell,
+    NotebookCellStatusBarAlignment,
+    NotebookCellStatusBarItem,
+    NotebookCellStatusBarItemProvider as VSCNotebookCellStatusBarItemProvider,
+    ProviderResult
+} from 'vscode';
+import { getCellStatusMessageBasedOnFirstCellErrorOutput } from './helpers/helpers';
+
+/**
+ * This class controls the status messages that we see on our Notebook cells
+ */
+@injectable()
+export class StatusBarProvider implements VSCNotebookCellStatusBarItemProvider {
+    // Weakmap, as we don't own the lifetime of either of these, just a helpful mapping for tests
+    private static cellStatusMappings = new WeakMap<NotebookCell, NotebookCellStatusBarItem>();
+
+    // Allow test code to see what status is currently on any cell
+    public static getCellStatusBarItem(cell: NotebookCell): NotebookCellStatusBarItem | undefined {
+        return StatusBarProvider.cellStatusMappings.get(cell);
+    }
+
+    // For any NotebookCell, check its output to see if we need to put up a status message
+    provideCellStatusBarItems(
```

Comment:
Won't we need to trigger an event letting VS Code know about potential changes?
* @rebornix 
Or will VS Code always call this method when ever there are changes to notebooks?

Instances when VS Code should call this:
* Execution has completed (outputs changed)
* Execution has gone into pending state (outputs changed)
* User clears outputs (outputs changed)

Basically when ever outputs change we need to ensure we either notify VS Code to call this or VS Code should call this when outputs change.

@IanMatthewHuff @rebornix  do we know if VS Code will call when outputs change?
If not then we'll need to trigger the event in this provider in all of the above cases.

Not sure about the scenario when user changes outputs (@rebornix do we have events when user cleas outputs?)
There are other places where we want to clea other metadata when user clears outputs.

Reply:
This isn't a showstopper, however we'd need to ensure this works as expected.

Reply:
@DonJayamanne What I did check was the following.

* Run a cell without an error (no status message)
* Create an error in the cell and run it (error status message show)
* Fix the error and run again (no status message)

I'll check for output clearing now.

Reply:
> Fix the error and run again (no status message)

Is this purely by chance of is this how VS Code works. I.e. does it query the methods when ever cell output or status changes? I'd prefer we know that for a fact instead of leaving this to chance.
@rebornix Please could you let us know when VS Code will query this method to get the latest status?

Reply:
It will be called

- When a cell begins or finishes executing (really when any internal metadata changes)
- When output changes
- When language changes
- When the cell scrolls into view
- When any provider fires `onDidChangeCellStatusBarItems`

I should have jsdoc that spells this out

Reply:
Awesome, then we don't need to do anything, thanks for the confirmation.
Easy simple powerful api.....